### ISSUE TITLE
read_block for DS1307

### ISSUE THREAD

**Issue by Henje** [lefticus/i2c_hacking#2]
After watching your stream I was reading through the kernel docs to see why the read_block does not work and I think I know what went wrong. I hope this is fine as an issue.

The main docs how the protocols are implemented in the kernel are in
https://www.kernel.org/doc/html/latest/i2c/i2c-protocol.html
https://www.kernel.org/doc/html/latest/i2c/smbus-protocol.html

The general i2c send/write operation is given by
`S Addr Wr [A] Data [A] Data [A] ... [A] Data [A] P`
which maps to the Slave Receiver Mode of the DS1307.

The general i2c recv/read operation is given by
`S Addr Rd [A] [Data] A [Data] A ... A [Data] NA P`
which maps to the Slave Transmitter Mode of the DS1307.

So all operations can be implemented with normal read() and write(). For the read_block one could do something like this
```cpp
std::uint8_t addr = 0;
write(handle, &addr, 1);
std::array<std::uint8_t, 32> buffer;
read(handle, buffer.data(), buffer.size());
```
The write is necessary to set the register pointer. This assumes the slave address is set using ioctl beforehand.

 A combined Slave Receive and Transmit of the DS1307 rewritten in the kernel docs format is
`S Addr Wr [A] Comm [A] S Addr Rd [A] [Data] A [Data] A ... A [Data] NA P`
which almost maps to the SMBus Block Read
`S Addr Wr [A] Comm [A] S Addr Rd [A] [Count] A [Data] A [Data] A ... A [Data] NA P`
So it seems they are incompatible, further all SMBus commands have to be implemented in the i2c adapter. For the pi 2 this is the bcm2835 driver. This driver implements all SMBus functions except Block Read and Block Process Call. This can be checked using `i2cdetect -F <bus>`.

The single byte write could also be changed to a i2c_smbus_write_byte, which is equivalent. Alternatively, one could combine the write and read with the ioctl transaction interface.

Caveat: I do not have access to the DS1307, I only tested the `i2cdetect -F` on my pi 2. This is only based on my reading of the docs.

**Comment by amozzhuhin** [comment]
Hi Henje.

Functions read() and write() works a bit differently. They generate one transaction on I2C bus. The address is taken from the current file descriptor position. So, to read or write from/to some address lseek() must be used to set address. More details can be seen in https://elixir.bootlin.com/linux/v5.10.11/source/drivers/i2c/i2c-dev.c#L116.

I have created pull request #3 with a multi byte implementation using i2c_smbus_read_i2c_block_data because libi2c is already used. But this may be also implemented also by read().

**Comment by amozzhuhin** [comment]
Ah, sorry, offset argument is not used in i2cdev_read. But as I can see it use I2C_M_RD flag without I2C_M_NOSTART and start new transaction every time. Looks like transaction with writing address and reading data can't be implemented in that way. Or I miss something?

**Comment by Henje** [comment]
Thanks for your comments. Yes my approach is not a single transaction. But my main goal was to achieve a single read operation to avoid the race condition Jason spoke about in his stream. I am pretty certain, that a write and a read will work, as I found the following code, which does the same: http://menetekel.e-technik.fh-muenchen.de/skripten/hardware/RasPi/Projekt-RTC/index.html

It seems I missed i2c_smbus_read_i2c_block_data which matches exactly the specification of the DS1307. But there is still the question of compatibility with the i2c adapter, nevertheless, it seems the read_i2c_block_data is available on the pi 2 while the read_block_data is not, weird. In any case, one could definitely use the ioctl interface to combine the write and read, but I see no big advantage here and the code is simpler, so your approach seems the best.

**Comment by Henje** [comment]
Fixed by 4333392
